File: _cscript/shared/PromiseProxy.cs
#if server
_ = require 'underscore'
require './Util'

smio = global.smoothio


#_as = (x) -> new smio.PromiseProxy x
#_as(@dbServer).connect(_as(mongodb.Db)).createCollection("_smio_sites", _as(mongodb.Collection))._callAll (err, col) ->


class smio.PromiseProxy
	constructor: (obj) ->
		@__smioprox =
			prot: if obj.prototype then obj.prototype else (if obj.constructor then obj.constructor.prototype else obj.__proto__)
			inst: if _.isFunction obj then null else obj
			chain: []
			parent: null
		template = if @__smioprox.prot then @__smioprox.prot else @__smioprox.inst
		makeProxyFunc = (that, name) ->
			() ->
				smio.logit "CALL #{name}"
				args = _.toArray arguments
				nextPromise = that
				if args and args.length and (lastArg = args[args.length - 1]) and lastArg['__smioprox'] and lastArg['_callAll']
					nextPromise = args[args.length - 1]
					nextPromise.__smioprox.parent = that
					args.length = args.length - 1
				that.__smioprox.chain.push [name, args, nextPromise]
				nextPromise
		for k, v of template
			if (_.isFunction v) and k isnt '_callAll'
				@[k] = makeProxyFunc @, k

	_callAll: (inst, cb, index) ->
		if (not cb) and (not index?) and inst and _.isFunction inst
			if @__smioprox.parent
				return @__smioprox.parent._callAll inst
			else
				cb = inst
				inst = @__smioprox.inst
		if not index?
			index = 0
		if not inst
			cb (new Error "No inst given!!"), null
		else
			run = (instance, name, args, next, rcb, ix, isNextDifferent) ->
				cbr = (err, ret) ->
					if err
						rcb err, ret
					else if (isNextDifferent and (ix < next.__smioprox.chain.length)) or ((not isNextDifferent) and (ix < (next.__smioprox.chain.length - 1)))
						next._callAll (if ret then ret else instance), rcb, if isNextDifferent then 0 else ix + 1
					else
						rcb null, ret
				instance[name].apply instance, smio.Util.Array.add args, cbr
			call = @__smioprox.chain[index]
			run inst, call[0], call[1], call[2], cb, index, call[2] isnt @
#endif

File: default/server/_jscript/shared/PromiseProxy.js
(function() {
  var smio, _;
  _ = require('underscore');
  require('./Util');
  smio = global.smoothio;
  smio.PromiseProxy = (function() {
    function PromiseProxy(obj) {
      var k, makeProxyFunc, template, v;
      this.__smioprox = {
        prot: obj.prototype ? obj.prototype : (obj.constructor ? obj.constructor.prototype : obj.__proto__),
        inst: _.isFunction(obj) ? null : obj,
        chain: [],
        parent: null
      };
      template = this.__smioprox.prot ? this.__smioprox.prot : this.__smioprox.inst;
      makeProxyFunc = function(that, name) {
        return function() {
          var args, lastArg, nextPromise;
          smio.logit("CALL " + name);
          args = _.toArray(arguments);
          nextPromise = that;
          if (args && args.length && (lastArg = args[args.length - 1]) && lastArg['__smioprox'] && lastArg['_callAll']) {
            nextPromise = args[args.length - 1];
            nextPromise.__smioprox.parent = that;
            args.length = args.length - 1;
          }
          that.__smioprox.chain.push([name, args, nextPromise]);
          return nextPromise;
        };
      };
      for (k in template) {
        v = template[k];
        if ((_.isFunction(v)) && k !== '_callAll') {
          this[k] = makeProxyFunc(this, k);
        }
      }
    }
    PromiseProxy.prototype._callAll = function(inst, cb, index) {
      var call, run;
      if ((!cb) && (!(index != null)) && inst && _.isFunction(inst)) {
        if (this.__smioprox.parent) {
          return this.__smioprox.parent._callAll(inst);
        } else {
          cb = inst;
          inst = this.__smioprox.inst;
        }
      }
      if (!(index != null)) {
        index = 0;
      }
      if (!inst) {
        return cb(new Error("No inst given!!"), null);
      } else {
        run = function(instance, name, args, next, rcb, ix, isNextDifferent) {
          var cbr;
          cbr = function(err, ret) {
            if (err) {
              return rcb(err, ret);
            } else if ((isNextDifferent && (ix < next.__smioprox.chain.length)) || ((!isNextDifferent) && (ix < (next.__smioprox.chain.length - 1)))) {
              return next._callAll((ret ? ret : instance), rcb, isNextDifferent ? 0 : ix + 1);
            } else {
              return rcb(null, ret);
            }
          };
          return instance[name].apply(instance, smio.Util.Array.add(args, cbr));
        };
        call = this.__smioprox.chain[index];
        return run(inst, call[0], call[1], call[2], cb, index, call[2] !== this);
      }
    };
    return PromiseProxy;
  })();
}).call(this);
